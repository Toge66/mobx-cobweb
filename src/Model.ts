/***************************************************
 * Created by nanyuantingfeng on 2020/6/2 12:53. *
 ***************************************************/
import { Model as _Model } from './datx'
import { withStorage } from './mixins/withStorage'
import { withNetActions } from './mixins/withNetActions'

/**** For DTS File (Don`t remove it) ****/
import { INetActionsMixinForModel } from './interfaces/INetActionsMixin'
import { IModelConstructor } from './datx'

export class Model extends withStorage(withNetActions(_Model)) {}
