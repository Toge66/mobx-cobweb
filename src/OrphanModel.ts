/***************************************************
 * Created by nanyuantingfeng on 2020/7/21 11:40. *
 ***************************************************/
import { Model } from './Model'
import { attribute } from './index'
import { ORPHAN_MODEL_ID_KEY, ORPHAN_MODEL_ID_VAL } from './helpers/consts'

export class OrphanModel extends Model {
  static enableAutoId = true

  static preprocess(data: any = {}) {
    data[ORPHAN_MODEL_ID_KEY] = ORPHAN_MODEL_ID_VAL
    return data
  }

  @attribute({ isIdentifier: true })
  public [ORPHAN_MODEL_ID_KEY]: string
}
