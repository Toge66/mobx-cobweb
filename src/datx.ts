/***************************************************
 * Created by nanyuantingfeng on 2020/7/28 09:28. *
 ***************************************************/
export * from '@issues-beta/datx'
