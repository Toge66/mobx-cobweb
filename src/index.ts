/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
export * from './interfaces'
export * from './helpers/model'

export { prop, View } from 'datx'

export type { IRawResponse } from './interfaces/IRawResponse'
export type { IRequestOptions } from './interfaces/IRequestOptions'

export { ResponseView } from './ResponseView'
export { GenericModel } from './GenericModel'
export { setNetworkAdapter } from './helpers/NetworkUtils'
export { clearAllCache, clearCacheByType } from './helpers/cache'

export { Collection } from './Collection'
export { Model } from './Model'

export { ListDataView } from './views/ListDataView'
export { TreeDataView } from './views/TreeDataView'
export { NetworkAdapter } from './adapter/NetworkAdapter'
