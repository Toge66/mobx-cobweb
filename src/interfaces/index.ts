/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 14:12. *
 ***************************************************/
export * from './types'
export * from './IRawResponse'
export * from './IRequestOptions'
export * from './IRawData'
export * from './IResponseHeaders'
export * from './INetworkAdapter'
