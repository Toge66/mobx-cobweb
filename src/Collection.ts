/***************************************************
 * Created by nanyuantingfeng on 2020/6/2 12:55. *
 ***************************************************/
import { Collection as _Collection, IModelConstructor, IType, PureModel, ICollectionConstructor } from 'datx'
import { withNetPatches } from './mixins/withNetPatches'
import { withStorage } from './mixins/withStorage'
import { INetPatchesCollectionMixin } from './interfaces/INetPatchesCollectionMixin'
import { IStorageCollectionMixin } from './interfaces/IStorageCollectionMixin'

const WithNetPatchesCollection: ICollectionConstructor<
  IStorageCollectionMixin<_Collection> & INetPatchesCollectionMixin<_Collection> & _Collection
> = withStorage(withNetPatches(_Collection))

export class Collection extends WithNetPatchesCollection {
  static register<T extends PureModel>(O: IModelConstructor<T>) {
    if (!this.types.find((Q) => Q.type === O.type)) {
      this.types.push(O)
    }
  }

  register<T extends PureModel>(O: IModelConstructor<T>) {
    ;((this.constructor as unknown) as Collection).register(O)
  }

  findOrphan<T extends PureModel>(model?: IType | IModelConstructor<T>): T {
    return this.findOne(model, '__$orphan_id__(0)')
  }
}
