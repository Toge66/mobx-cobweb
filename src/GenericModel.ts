/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
import { IModelConstructor, PureModel } from 'datx'
import { withNetActions } from './mixins/withNetActions'
import { INetActionsMixin } from './interfaces/INetActionsMixin'
const WithNetActionModel: IModelConstructor<INetActionsMixin<PureModel> & PureModel> = withNetActions(PureModel)

export class GenericModel extends WithNetActionModel {}
