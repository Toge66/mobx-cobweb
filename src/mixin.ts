/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
import {
  ICollectionConstructor,
  IModelConstructor,
  isCollection,
  isModel,
  isView,
  IViewConstructor,
  PureCollection,
  PureModel,
  View
} from 'datx'

import { decorateCollection } from './implements/decorateCollection'
import { decorateModel } from './implements/decorateModel'
import { decorateView } from './implements/decorateView'
import { ISkeletonCollection, ISkeletonModel, ISkeletonModelConstructor } from './interfaces'

export function skeleton<T extends PureModel>(
  BaseClass: IModelConstructor<T>
): ISkeletonModelConstructor<T & ISkeletonModel>

export function skeleton<T extends PureCollection>(
  BaseClass: ICollectionConstructor<T>
): ICollectionConstructor<T & ISkeletonCollection>

export function skeleton<T extends PureModel | PureCollection | View>(
  BaseClass: IModelConstructor<T> | ICollectionConstructor<T> | IViewConstructor<T>
) {
  if (isModel(BaseClass)) {
    return decorateModel(BaseClass as typeof PureModel)
  } else if (isCollection(BaseClass)) {
    return decorateCollection(BaseClass as typeof PureCollection)
  } else if (isView(BaseClass)) {
    return decorateView<T>(BaseClass as typeof View)
  }

  throw new Error('The instance needs to be a model, collection or a view')
}
