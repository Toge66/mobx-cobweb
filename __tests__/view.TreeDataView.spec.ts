/***************************************************
 * Created by nanyuantingfeng on 2020/6/2 16:42. *
 ***************************************************/
import { collection, useFixturesByGET } from './config'
import { TreeDataView } from '../src'
import TreeDataDEMOModel from './models/TreeDataDEMOModel'

describe('TreeDataView', () => {
  let scope: any = null

  beforeEach(() => {
    scope = useFixturesByGET()
  })

  it('should be basic apis', async () => {
    scope.get(TreeDataDEMOModel.endpoint + '/roots').reply(200, (uri: string) => {
      return JSON.stringify({
        items: [
          {
            id: 0,
            name: '0'
          }
        ]
      })
    })

    const treeDataView = new TreeDataView<TreeDataDEMOModel>(TreeDataDEMOModel, collection)
    await treeDataView.searchRoots()
    expect(treeDataView.data.length).toBe(1)
    const root = treeDataView.data[0]
    expect(collection.findAll(TreeDataDEMOModel).length).toBe(1)

    scope
      .persist()
      .post(TreeDataDEMOModel.endpoint + '/$0/children')
      .reply(200, (uri: string, body: any) => {
        const { limit } = body
        const { start, count } = limit
        const oo = []
        let i = -1
        while (++i < count) {
          oo.push({ id: start + i, name: start + i })
        }
        return JSON.stringify({ items: oo, count: 100 })
      })

    await treeDataView.infinite(root, { selector: { limit: [0, 30] } })
    const childrenList = treeDataView.getInfiniteListDataView(root)
    expect(treeDataView.data.length).toBe(1)
    expect(root.children.length).toBe(30)
    await childrenList.next()
    expect(root.children.length).toBe(60)
    expect(collection.findAll(TreeDataDEMOModel).length).toBe(60)

    scope
      .persist()
      .get(TreeDataDEMOModel.endpoint + '/$20/parents')
      .reply(200, (uri: string) => {
        return JSON.stringify({ items: [{ id: 200 }, { id: 10 }, { id: 0 }] })
      })

    const resp = await treeDataView.searchParents(20)

    expect(resp.data.length).toBe(3)
  })
})
