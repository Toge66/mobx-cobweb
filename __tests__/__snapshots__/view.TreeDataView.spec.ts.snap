// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeDataView should be fetch children nodes 1`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "order",
      "id",
    ],
    "id": "6Rk9l1WYNM0400:72092047",
    "refs": Object {
      "children": Object {
        "model": "organization.Department",
        "type": 1,
      },
      "parentId": Object {
        "model": "organization.Department",
        "type": 0,
      },
    },
    "type": "organization.Department",
  },
  "active": true,
  "children": Array [],
  "code": "",
  "corporationId": "6Rk9l1WYNM0400",
  "createTime": 1574211989469,
  "id": "6Rk9l1WYNM0400:72092047",
  "name": "线上支持组",
  "nameSpell": "XIANSHANGZHICHIZU",
  "order": 72092047,
  "parentId": "6Rk9l1WYNM0400:50128174",
  "updateTime": 1574211989469,
  "version": 1,
}
`;

exports[`TreeDataView should be fetch parent node 1`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "order",
      "id",
    ],
    "id": "6Rk9l1WYNM0400:50128174",
    "refs": Object {
      "children": Object {
        "model": "organization.Department",
        "type": 1,
      },
      "parentId": Object {
        "model": "organization.Department",
        "type": 0,
      },
    },
    "type": "organization.Department",
  },
  "active": true,
  "children": Array [
    "6Rk9l1WYNM0400:72167029",
  ],
  "code": "",
  "corporationId": "6Rk9l1WYNM0400",
  "createTime": 1574211989469,
  "id": "6Rk9l1WYNM0400:50128174",
  "name": "测试支持大部3",
  "nameSpell": "CESHIZHICHIDABU3",
  "order": 2,
  "parentId": "6Rk9l1WYNM0400:49525176",
  "updateTime": 1574211989469,
  "version": 1,
}
`;

exports[`TreeDataView should be fetched children nodes 1`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "order",
      "id",
    ],
    "id": "6Rk9l1WYNM0400:72092047",
    "refs": Object {
      "children": Object {
        "model": "organization.Department",
        "type": 1,
      },
      "parentId": Object {
        "model": "organization.Department",
        "type": 0,
      },
    },
    "type": "organization.Department",
  },
  "active": true,
  "children": Array [],
  "code": "",
  "corporationId": "6Rk9l1WYNM0400",
  "createTime": 1574211989469,
  "id": "6Rk9l1WYNM0400:72092047",
  "name": "线上支持组",
  "nameSpell": "XIANSHANGZHICHIZU",
  "order": 72092047,
  "parentId": "6Rk9l1WYNM0400:50128174",
  "updateTime": 1574211989469,
  "version": 1,
}
`;

exports[`TreeDataView should be fetched parent node 1`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "order",
      "id",
    ],
    "id": "6Rk9l1WYNM0400:50128174",
    "refs": Object {
      "children": Object {
        "model": "organization.Department",
        "type": 1,
      },
      "parentId": Object {
        "model": "organization.Department",
        "type": 0,
      },
    },
    "type": "organization.Department",
  },
  "active": true,
  "children": Array [
    "6Rk9l1WYNM0400:72167029",
  ],
  "code": "",
  "corporationId": "6Rk9l1WYNM0400",
  "createTime": 1574211989469,
  "id": "6Rk9l1WYNM0400:50128174",
  "name": "测试支持大部3",
  "nameSpell": "CESHIZHICHIDABU3",
  "order": 2,
  "parentId": "6Rk9l1WYNM0400:49525176",
  "updateTime": 1574211989469,
  "version": 1,
}
`;
